sixtyfps_runtime/corelib/graphics/image.rs

```diff
@@ -11,6 +11,226 @@ LICENSE END */
+///     // render beautiful circle or other shapes here
+/// }
+///
+/// let mut pixel_buffer = SharedPixelBuffer::<rgb::RGB8>::new(320, 200);
+///
+/// low_level_render(pixel_buffer.width(), pixel_buffer.height(),
+///                  pixel_buffer.as_mut_slice().as_bytes_mut());
+///
+/// let image: Image = SharedImageBuffer::RGB8(pixel_buffer).into();
+/// ```
+///
+/// Another use-case is to import existing image data into SixtyFPS, by
+/// creating a new SharedImageBuffer through cloning of another image
+/// type.
+///
+/// The following example uses the popular [image crate](https://docs.rs/image/) to
+/// load a `.png` file from disk, apply brightening filter on it and then import
+/// it into an `[Image]`:
+/// ```no_run
+/// # use sixtyfps_corelib::graphics::{SharedPixelBuffer, SharedImageBuffer, Image};
+/// let mut cat_image = image::open("cat.png").expect("Error loading cat image").into_rgba8();
+///
+/// image::imageops::colorops::brighten_in_place(&mut cat_image, 20);
+///
+/// let buffer = SharedImageBuffer::RGBA8(SharedPixelBuffer::clone_from_slice(
+///     cat_image.as_raw(),
+///     cat_image.width() as _,
+///     cat_image.height() as _,
+/// ));
+/// let image: Image = buffer.into();
+/// ```
+///
+
+#[derive(Clone, Debug)]
+pub enum SharedImageBuffer {
+    /// This variant holds the data for an image where each pixel has three color channels (red, green,
+    /// and blue) and each channel is encoded as unsigned byte.
+    RGB8(SharedPixelBuffer<rgb::RGB8>),
+    /// This variant holds the data for an image where each pixel has four color channels (red, green,
+    /// blue and alpha) and each channel is encoded as unsigned byte.
+    RGBA8(SharedPixelBuffer<rgb::RGBA8>),
+    /// This variant holds the data for an image where each pixel has four color channels (red, green,
+    /// blue and alpha) and each channel is encoded as unsigned byte. In contrast to [`Self::RGBA8`],
+    /// this variant assumes that the alpha channel is also already multiplied to each red, green and blue
+    /// component of each pixel.
+    /// Only construct this format if you know that your pixels are encoded this way. It is more efficient
+    /// for rendering.
+    RGBA8Premultiplied(SharedPixelBuffer<rgb::RGBA8>),
+}
+
+impl SharedImageBuffer {
+    /// Returns the width of the image in pixels.
+    #[inline]
+    pub fn width(&self) -> usize {
+        match self {
+            Self::RGB8(buffer) => buffer.width(),
+            Self::RGBA8(buffer) => buffer.width(),
+            Self::RGBA8Premultiplied(buffer) => buffer.width(),
+        }
+    }
+
+    /// Returns the height of the image in pixels.
+    #[inline]
+    pub fn height(&self) -> usize {
+        match self {
+            Self::RGB8(buffer) => buffer.height(),
+            Self::RGBA8(buffer) => buffer.height(),
+            Self::RGBA8Premultiplied(buffer) => buffer.height(),
+        }
+    }
+}
+
+impl PartialEq for SharedImageBuffer {
```

Comment:
Why Is the built-in derive not working?

Reply:
Initially that type contained an `ImgVec`, which didn't implement `PartialEq`. Now we *could* use `PartialEq` from the built-in derive, but it would mean `SharedVector`'s `PartialEq`, which will compare pixel by pixel.

For a `property <image>` that's rather expensive, hence this implementation that is guaranteed to return false if the images differ, but it won't return true for all equal images.